Add mission GDI 8b

reaperrr opened the pull request:
Probably still a bit rough around the edges, but this mission has some tricky aspects so I expect this to need a few review passes anyway. Some early feedback could help get there faster.

fruestueck committed:
Add GDI 8b

Smittytron commented on mods/cnc/maps/gdi08b/import.log.txt:
```diff
@@ -0,0 +1,21 @@
+Import log, fruehstueck
```
Stowaway file

Smittytron commented on mods/cnc/maps/gdi08b/scg08eb.ini:
```diff
@@ -0,0 +1,64 @@
+[Triggers]
```
Stowaway file

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b-AI.lua:
```diff
@@ -0,0 +1,249 @@
+
+StartAI = function()
+	Nod.Cash = StartingCash
+	--GuardBase()
```
Commented out

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b-AI.lua:
```diff
@@ -0,0 +1,249 @@
+end)
+
+-- units destroyed, send heli, eg. harv, tnk, bggy,...
+Trigger.OnKilled(harv1, function()
```
Should be able to combine all these into a table and use
```
Utils.Do(NodUnits, function(unit)
	Trigger.OnKilled(unit, function()
		SendHeli(NodHelis[2])
```

reaperrr replied:
Before I update the PR, I've been wondering: How does lua count? Is `NodHelis[2]` correct (1,2,3), or should it be `NodHelis[1]` (0,1,2)?

Smittytron replied:
It counts 1,2,3

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,199 @@
+--[[
+   Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
2021

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,199 @@
+
+TerrorHeliDelay = { hard = 130, normal = 170, easy = 210 }
+BaseHeliDelay = { hard = 100, normal = 130, easy = 160 }
+NodHelis = {
```
For consistency I'd put `{` on its own line.

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,199 @@
+	
+	Trigger.AfterDelay(DateTime.Minutes(1), function() SendWaves(1, AutoAttackWaves) end)
+	--Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceInfantry(handofnod) end) -- DateTime.Minutes(3)
+	--Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceVehicle(nodairfield) end) -- DateTime.Minutes(3)
```
Commented out

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,199 @@
+	--Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceInfantry(handofnod) end) -- DateTime.Minutes(3)
+	--Trigger.AfterDelay(DateTime.Minutes(3), function() ProduceVehicle(nodairfield) end) -- DateTime.Minutes(3)
+	
+	local InitialArrivingUnits = { 
```
New line for `{`

Smittytron commented on mods/cnc/maps/gdi08b/map.yaml:
```diff
@@ -0,0 +1,1049 @@
+Actors:
+	Actor0: wood
+		Location: 10,62
+		Owner: Neutral
```
Walls need owners.

Smittytron commented on mods/cnc/maps/gdi08b/rules.yaml:
```diff
@@ -0,0 +1,170 @@
+	Buildable:
+		Prerequisites: ~disabled
+		
+GTWR:
```
Add space here and for `GUN`

reaperrr commented:
Updated.

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,208 @@
+		GDI.MarkFailedObjective(ProtectHospital)
+	end)
+	
+	ProtectCivilians = GDI.AddObjective("Protect the Civilians.")
```
This would be more informative if it displayed the number of civilians required per difficulty level.

Smittytron reviewed (dismissed):
The civilians may wander on to tiberium and die. I would either override`Wanders `for this mission or assign the civilians to player control. 

Got this crash at the 3 minute mark.
```
Fatal Lua Error: [string "-- Make copies of global functions to preve..."]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at OpenRA.Exts.Random[T](IEnumerable`1 ts, MersenneTwister r, Boolean throws) in C:\Users\Smitty\OpenRA\OpenRA.Game\Exts.cs:line 162
   at OpenRA.Mods.Common.Scripting.UtilsGlobal.Random(LuaValue[] collection) in C:\Users\Smitty\OpenRA\OpenRA.Mods.Common\Scripting\Global\UtilsGlobal.cs:line 91
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OpenRA.Scripting.ScriptMemberWrapper.Invoke(LuaVararg args) in C:\Users\Smitty\OpenRA\OpenRA.Game\Scripting\ScriptMemberWrapper.cs:line 75
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Eluant.LuaRuntime.MethodWrapper.Invoke(Object[] parms)
   at Eluant.LuaRuntime.MakeManagedCall(IntPtr state, MethodWrapper wrapper)
   at OpenRA.Scripting.ScriptContext.FatalError(String message)
   at OpenRA.Mods.Common.Scripting.TriggerGlobal.<>c__DisplayClass2_0.<AfterDelay>b__0()
   at OpenRA.Effects.DelayedAction.<Tick>b__3_0(World w)
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)
```

reaperrr commented:
> The civilians may wander on to tiberium and die. I would either override Wanders for this mission or assign the civilians to player control.

They shouldn't!
> AvoidTerrainTypes: Tiberium, BlueTiberium

Is set on the civ default in TD, for both `Wanders` and `ScaredyCat`.

If they still wander onto tiberium, it means the `AvoidTerrainTypes` logic is broken, that would be something for another PR then.

reaperrr commented:
Did some testing. The avoid terrain logic works, it only occasionally happens that tiberium either spawns on a cell a civilian stands on, or spawns on a cell a civilian already started wandering into. At least in my tests it didn't happen often enough to warrant extra effort to work around this.

Also, the problem with making the civilians player-controllable would be that it would allow players to move all of them into the base, making their protection almost too easy (relatively speaking, compared to the original).

I think the bigger problem is that the civilians often block your vehicles from moving through the village, I'll make some changes to tone down at least that problem.

Smittytron commented:
I'm guessing they took some area damage in a fight and I thought it was from tiberium.

reaperrr commented:
Fixed that the map selection image showed the arrows of both 08 missions.
Fixed crash and revamped production/auto-attacking. Tested this myself and apart from fine-tuning the difficulty, this should work well enough now.
Reduced civilian wander frequency and increased health for this mission. Protecting the village with sandbags is still recommended, though.
Made the 'protect civilians' objective show how many you need to keep alive.

Made the village a little easier to navigate through, by
- tweaking the initial placement of some civilians
- moving the top-right civ buildings up by a cell
- replacing the original crossroad with a variant that doesn't have an impassable ruin to the side

Tweaked some attack delays, but since I'm a poor player I can't judge if the higher difficulties are manageable.

reaperrr commented:
Updated again, enabled 'Civilian killed' notification for this mission.

Smittytron commented on mods/cnc/maps/gdi08b/gdi08b.lua:
```diff
@@ -0,0 +1,207 @@
+   the License, or (at your option) any later version. For more
+   information, see COPYING.
+]]
+
```
Will need `Difficulty = Map.LobbyOption("difficulty")` added because of #19426.

Smittytron commented:
LGTM otherwise.

reaperrr commented:
Updated (and rebased just in case).

dtluna commented:
@reaperrr Played through the mission and I believe Moebius can be selected with Q which is not good :sweat_smile:

dtluna commented on mods/cnc/maps/gdi08b/rules.yaml:
```diff
@@ -0,0 +1,180 @@
+		StartingMusic: march
+		VictoryMusic: gdi_win1
+	MissionData:
+		Briefing: Dr. Moebius is establishing a Hospital to treat the civilians in the region who are falling ill due to exposure to Tiberium.\n\nProtect Moebius and the civilians.\n\nEliminate Nod presence in the area.
```
I think you should replace "Moebius" with "Mobius" because the mission is called "Protect **Mobius** In Bratislava", which is the canon spelling.

reaperrr committed:
Fix unarmed civilians being selected as combat units

reaperrr commented:
Updated.

- Used Mobius spelling in briefing
- Changed unarmed civilians to be considered support units so they (including Mobius) don't get selected by Q
- Moved Chinook entry point to upper right map border, rather than middle of the map

dtluna commented:
@reaperrr I'm good with this now

Smittytron commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d459e0aebbdbda9df60514edaa69200fea03f344
